// solidity-docgen/dist/solc.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ava_1 = __importDefault(require("ava"));
const solc_1 = require("./solc");
function smokeTest(t, adapter, version) {
    const output = adapter.compile({
        language: 'Solidity',
        sources: {
            test: {
                content: `pragma solidity ^${version}; contract Foo { }`,
            },
        },
        settings: {
            outputSelection: solc_1.outputSelection,
        },
    });
    t.is('object', typeof output);
    t.is(undefined, output.errors);
}
ava_1.default('smoke test 0.6', async (t) => {
    const adapter = await solc_1.SolcAdapter.require('solc');
    smokeTest(t, adapter, '0.6');
});
ava_1.default('smoke test 0.5', async (t) => {
    const adapter = await solc_1.SolcAdapter.require('solc-0-5');
    smokeTest(t, adapter, '0.5');
});
ava_1.default('smoke test 0.4', async (t) => {
    const adapter = await solc_1.SolcAdapter.require('solc-0-4');
    smokeTest(t, adapter, '0.4');
});
